### PR TITLE
Add continue and break statement support

### LINKED ISSUE #22321: Add continue and break statement support to the new parser
https://ballerina.io/spec/lang/2020R1/#section_7.22
https://ballerina.io/spec/lang/2020R1/#section_7.23

### PR DESCRIPTION
**Pull request by rdulmina**
## Purpose
Add continue and break statement support for the new parser

Fixes #22321

## Approach

1. Add new tokens.
2. Add continue and break statement support.
3. Add recovery logic for both statements.
4. Add Test cases to test new statements

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/tree/STNodeFactory.java [hunk 1]
```diff
@@ -335,5 +346,7 @@ public static STNode createVersionPart(STNode leadingDot, STNode versionNumber)
 
     public static STNode createImportPrefix(STNode asKeyword, STNode prefix) {
         return new STImportPrefix(asKeyword, prefix);
+
     }
-}
+
+}
```
Comment (SupunS) [inline compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/tree/STNodeFactory.java:352]: add a newline at EOF

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/tree/SyntaxKind.java [hunk 2]
```diff
@@ -50,8 +50,11 @@
     WHILE_KEYWORD(207, "while"),
     CHECK_KEYWORD(208, "check"),
     CHECKPANIC_KEYWORD(209, "checkpanic"),
+    CONTINUE_KEYWORD(212,"continue"),
+    BREAK_KEYWORD(213,"break"),
```
Comment (SupunS) [inline compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/tree/SyntaxKind.java:54]: seems this is not in the correct order. Also next number should be 211

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/tree/SyntaxKind.java [hunk 3]
```diff
@@ -110,6 +113,10 @@
     WHILE_STATEMENT(1205),
     CALL_STATEMENT(1206),
     PANIC_STATEMENT(1207),
+    CONTINUE_STATEMENT(1209),
+    BREAK_STATEMENT(1210),
```
Comment (SupunS) [inline compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/tree/SyntaxKind.java:117]: next number should be 1208

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java [hunk 4]
```diff
@@ -3486,4 +3495,62 @@ private STNode parseCheckingKeyword() {
...
+    private STNode parseContinueStatement() {
+        startContext(ParserRuleContext.CONTINUE_STATEMENT);
+        STNode continueKeyword = parseContinueKeyword();
+        STNode semicolon = parseSemicolon();
+        endContext();
+        return STNodeFactory.createContinueStatement(continueKeyword, semicolon);
+    }
+
+    /**
+     * Parse Continue-keyword.
```
Comment (SupunS) [inline compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java:3514]: Use 'sentence-case' for comments

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java [hunk 5]
```diff
@@ -3486,4 +3495,62 @@ private STNode parseCheckingKeyword() {
...
+    private STNode parseBreakStatement() {
+        startContext(ParserRuleContext.BREAK_STATEMENT);
+        STNode breakKeyword = parseBreakKeyword();
+        STNode semicolon = parseSemicolon();
+        endContext();
+        return STNodeFactory.createBreakStatement(breakKeyword, semicolon);
+    }
+
+    /**
+     * Parse Continue-keyword.
```
Comment (SupunS) [inline compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java:3543]: Incorrect doc comment

**Comment by SupunS** [comment]
@rdulmina There are some conflicts. Can you please resolve them?

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/tree/SyntaxKind.java (hunk 3 above)
Comment (rdulmina) [reply compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/tree/SyntaxKind.java:117]: return statement has used 1208 supun should I use it anyway?

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/tree/SyntaxKind.java (hunk 2 above)
Comment (rdulmina) [reply compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/tree/SyntaxKind.java:54]: I'll fix that

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/tree/SyntaxKind.java (hunk 3 above)
Comment (SupunS) [reply compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/tree/SyntaxKind.java:117]: Use the next immediate available number.

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/syntax/tree/BreakStatement.java [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+        breakToken = createToken(0);
+        return breakToken;
+    }
+
+    public Token semicolonToken() {
+        if (semicolonToken != null) {
+            return semicolonToken;
+        }
+
+        semicolonToken = createToken(2);
```
Comment (SupunS) [inline compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/syntax/tree/BreakStatement.java:44]: Shouldn't this be `createToken(1);`?

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/syntax/tree/ContinueStatement.java [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+        continueToken = createToken(0);
+        return continueToken;
+    }
+
+    public Token semicolonToken() {
+        if (semicolonToken != null) {
+            return semicolonToken;
+        }
+
+        semicolonToken = createToken(2);
```
Comment (SupunS) [inline compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/syntax/tree/ContinueStatement.java:44]: Shouldn't this be `createToken(1);`?

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java [hunk 8]
```diff
@@ -3486,4 +3495,62 @@ private STNode parseCheckingKeyword() {
...
+    private STNode parseBreakKeyword() {
+        STToken token = peek();
+        if (token.kind == SyntaxKind.BREAK_KEYWORD) {
+            return consume();
+        } else {
+            Solution sol = recover(token, ParserRuleContext.BREAK_KEYWORD);
+            return sol.recoveredNode;
+        }
+    }
+}
```
[the remark this answers is not in the document]
Comment (rdulmina) [reply compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java]: done

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/tree/STBreakStatement.java [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+import io.ballerinalang.compiler.syntax.tree.NonTerminalNode;
+import io.ballerinalang.compiler.syntax.tree.BreakStatement;
+
+public class STBreakStatement extends STStatement{
+    public final STNode breakKeyword;
+    public final STNode semicolonToken;
+
+
+    STBreakStatement(STNode breakKeyword,
+                        STNode semicolonToken) {
```
[the remark this answers is not in the document]
Comment (rdulmina) [reply compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/tree/STBreakStatement.java]: corrected

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/tree/STBreakStatement.java [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+        this.bucketCount = 2;
+        this.childBuckets = new STNode[this.bucketCount];
+        this.addChildNode(breakKeyword, 0);
+        this.addChildNode(semicolonToken, 1);
+
+    }
+
+    @Override
+    public Node createFacade(int position, NonTerminalNode parent) {
+
```
[the remark this answers is not in the document]
Comment (rdulmina) [reply compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/tree/STBreakStatement.java]: done

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/tree/STNodeFactory.java (hunk 1 above)
Comment (rdulmina) [reply compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/tree/STNodeFactory.java]: done

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/tree/SyntaxKind.java (hunk 2 above)
Comment (rdulmina) [reply compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/tree/SyntaxKind.java:55]: done

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java (hunk 4 above)
Comment (rdulmina) [reply compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java]: done

**Comment by rdulmina** [comment]
@SupunS I have resolved the conflicts. Can you review again

File: compiler/ballerina-parser/src/test/resources/statements/break-stmt/break_stmt_assert_1.json [hunk 11]
```diff
@@ -0,0 +1,11 @@
...
+  "kind": "BREAK_STATEMENT",
+  "children": [
+    {
+      "kind": "BREAK_KEYWORD"
+    },
+    {
+      "kind": "SEMICOLON_TOKEN"
+    }
+  ]
+}
```
Comment (sameerajayasoma) [inline compiler/ballerina-parser/src/test/resources/statements/break-stmt/break_stmt_assert_1.json:11]: Can you dd a new line here?

File: compiler/ballerina-parser/src/test/resources/statements/break-stmt/break_stmt_assert_2.json [hunk 12]
```diff
@@ -0,0 +1,12 @@
...
+  "children": [
+    {
+      "kind": "BREAK_KEYWORD"
+    },
+    {
+      "kind": "SEMICOLON_TOKEN",
+      "isMissing": true
+    }
+  ]
+}
```
Comment (sameerajayasoma) [inline compiler/ballerina-parser/src/test/resources/statements/break-stmt/break_stmt_assert_2.json:12]: Can you dd a new line here?

File: compiler/ballerina-parser/src/test/resources/statements/continue-stmt/continue_stmt_assert_1.json [hunk 13]
```diff
@@ -0,0 +1,11 @@
...
+  "kind": "CONTINUE_STATEMENT",
+  "children": [
+    {
+      "kind": "CONTINUE_KEYWORD"
+    },
+    {
+      "kind": "SEMICOLON_TOKEN"
+    }
+  ]
+}
```
Comment (sameerajayasoma) [inline compiler/ballerina-parser/src/test/resources/statements/continue-stmt/continue_stmt_assert_1.json:11]: Can you dd a new line here?

File: compiler/ballerina-parser/src/test/resources/statements/continue-stmt/continue_stmt_assert_2.json [hunk 14]
```diff
@@ -0,0 +1,12 @@
...
+  "children": [
+    {
+      "kind": "CONTINUE_KEYWORD"
+    },
+    {
+      "kind": "SEMICOLON_TOKEN",
+      "isMissing": true
+    }
+  ]
+}
```
Comment (sameerajayasoma) [inline compiler/ballerina-parser/src/test/resources/statements/continue-stmt/continue_stmt_assert_2.json:12]: Can you dd a new line here?

Comment (rdulmina) [reply compiler/ballerina-parser/src/test/resources/statements/continue-stmt/continue_stmt_assert_2.json]: Done @sameerajayasoma

**Pull request merged**